### PR TITLE
2410: Introduce process killing finally block in GitCommits

### PR DESCRIPTION
**Pull request by zhaosongzs**
Recently, I found that for pr-external container, in GitCommits, the exit of the git process could time out and then the process will be stuck in sleeping state and eat memory. With Pavel's suggestion, I think we should introduce the process killing finally block.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-2410](https://bugs.openjdk.org/browse/SKARA-2410): Introduce process killing finally block in GitCommits (**Bug** - P2)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1697/head:pull/1697` \
`$ git checkout pull/1697`

Update a local copy of the PR: \
`$ git checkout pull/1697` \
`$ git pull https://git.openjdk.org/skara.git pull/1697/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1697`

View PR using the GUI difftool: \
`$ git pr show -t 1697`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1697.diff">https://git.openjdk.org/skara/pull/1697.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1697#issuecomment-2486509886)
</details>

### CODE REVIEW

**Comment by zhaosongzs** [comment]
Since most of the code are from Pavel.  I am going to set him as the author
/author @pavelrappo

**Review by erikj79 — APPROVED** [review]

**Comment by zhaosongzs** [comment]
Thank you for the review!
/integrate

File: vcs/src/main/java/org/openjdk/skara/vcs/git/GitCommits.java [hunk 1]
```diff
@@ -133,23 +138,38 @@ public void close() throws IOException {
...
                         throw new IOException("'" + String.join(" ", command) + "' exited with code: " + exitCode);
                     }
                 }
-            } catch (InterruptedException e) {
-                throw new IOException("'" + String.join(" ", command) + "' was interrupted", e);
+            }
+        } catch (InterruptedException e) {
+            throw new IOException("'" + String.join(" ", command) + "' was interrupted", e);
+        } finally {
+            if (p.isAlive()) {
```
[the remark this answers is not in the document]
Comment (erikj79) [reply vcs/src/main/java/org/openjdk/skara/vcs/git/GitCommits.java:170]: Oh, you are right. I didn't look at the whole class.

[the remark this answers is not in the document]
Comment (zhaosongzs) [reply vcs/src/main/java/org/openjdk/skara/vcs/git/GitCommits.java:170]: Do you mean we need to iterate all the processes and destroy them in the finally block?

[the remark this answers is not in the document]
Comment (erikj79) [reply vcs/src/main/java/org/openjdk/skara/vcs/git/GitCommits.java:170]: I think the loop needs to make sure the inner close method is called on each process by catching exceptions and save one of them, probably the first one, and then rethrow it once all processes have been dealt with. In practice I doubt there will ever be multiple processes though. That would only happen if `::iterator` is called multiple times.

[the remark this answers is not in the document]
Comment (pavelrappo) [reply vcs/src/main/java/org/openjdk/skara/vcs/git/GitCommits.java:170]: > I think the loop needs to make sure the inner close method is called on each process by catching exceptions

Correct.

> and saving one of them, probably the first one, and then rethrow it once all processes have been dealt with.

You don't need to figure out which exception you need to save. Save all of them. For that, use java.lang.Throwable::addSuppressed.

> In practice I doubt there will ever be multiple processes though. That would only happen if `::iterator` is called multiple times.

Also true. I only pointed out that the `close` method is not adequate for the current state of the class. Either `close` should account for multiple iterators, or it should throw an exception if more than one iterator is created.

I guess `GitCommits` is itera_BLE_ and not itera_TOR_ only to be able to be used with the enhanced `for` statement. Tradeoffs were made.

**Pull request closed without merging**